rfcs/0027-frontier-synchronization.md

```diff
@@ -0,0 +1,177 @@
+
+IMPLEMENTATION NOTE: The entire chain of mask destruction hooks needs to be executed synchronously within a single async cycle in order to update breadcrumb subtrees without fear of race conditions.
+
+#### Frontier Read Monad
+
+A new monad is introduced for specifying reads from the transition frontier (see [alternatives section](#alternatives-misread) for explanation on why). All existing read functions on the transition frontier will instead by turned into functions which do not take in a transition frontier and return a result wrapped in the `'a Transition_frontier.Read.t` monad. The monad is designed to interact with deferreds so that async programming is still accessible during reads when necessary.
+
+#### New Frontier Interface
+
+Below is a partial signature for what the new transition frontier interface would look like. Note that all base read/write functions from the transition frontier are removed, and instead, only the read monad and write actions are accessible. The only way to interact with the transition frontier thus becomes the `read` and `write` functions.
+
+```ocaml
+module Transition_frontier : sig
+  module Breadcrumb : sig
+    module Full : sig
+      type t
+      val block : t -> Block.Validated.t
+      val staged_ledger : t -> Transaction_snark_scan_state.t
+    end
+
+    module Partial : sig
+      type t
+      val block : t -> Block.Validated.t
+      val scan_state : t -> Transaction_snark_scan_state.t
+    end
+
+    type full
+    type partial
+
+    type 'typ t =
+      | Full : Full.t -> full t
+      | Partial : Partial.t -> partial t
+    type 'typ breadcrumb = 'typ t
+
+    val external_transition : t -> External_transition.Validated.t
+
+    module Any : sig
+      type t = T : _ breadcrumb -> t
+      type ref = t ref
+      val wrap : _ breadcrumb -> t
+      val external_transition : t -> External_transition.Validated.t
+      val staged_state :
+           t
+        -> [ `Staged_ledger of Staged_ledger.t
+           | `Scan_state of Transaction_snark_scan_state.t ]
+    end
+  end
+
+  module Extensions : sig
+    type t
+
+    (* ... *)
+  end
+
+  module Read : sig
+    include Monad.S
+
+    (* to wrap computations in the monad *)
+    val deferred : 'a Deferred.t -> 'a t
+
+    val find : State_hash.t -> Breadcrumb.Any.ref option t
+
+    (* necessary extension access will be provided here, Extensions.t will not be directly accessible here *)
+
+    (* ... *)
+  end
+
+  type t
+
+  (* Actions are representations of mutations (writes) to perform on
+   * the frontier. There is only one action that can be performed on
+   * a frontier right now, adding breadcrumbs, but we can add either
+   * a single breadcrumb, or a subtree of breadcrumbs. Future actions
+   * can also be added as needed *)
+  type action =
+    | Add_breadcrumb of Breadcrumb.Full.t
+    | Add_subtree of Breadcrumb.Full.t Rose_tree.t
+
+  val read :
+       t
+    -> f:('a Read.t)
+    -> 'a Deferred.t
```

Comment:
You can exfiltrate the refs with `read frontier (Read.find hash)`, and then use them without the synchronization guarantee we're trying to establish. I'm not sure if the trick Haskell's ST monad uses works in Ocaml, but if it does that'll prevent this.

Reply:
This was a mistake in how I wrote up the pseudo code signature. I have updated the signature in a way that addresses this issue.